### DISCUSSION TITLE
Are there any guidelines to speed up network connection by choosing the right curl opt headers with libcurl

### DISCUSSION THREAD

**Discussion by ganeshkamath89** [curl/curl#6761] [Q&A] [answered]
I am using libcurl to connect from my client to servers (internal and 3rd party). 

While choosing curl opt headers :

1. How do I choose the headers if my priority is speed.
2. How do I choose headers if my priority is lesser TCP Retransmission / TCP ACK.
3. How do I choose the headers if I prefer more persistant communication by sacrificing speed.

If you can share examples of changing curl opt headers that have made a real difference in user experience, please share.

**Comment by jay** [comment, marked as answer]
Are you having performance issues? curl should perform well with default settings.

**Reply by ganeshkamath89** [reply]
Hi @jay. I have observed the following in Wireshark logs:
1) When I connect using client with Curl, the Wireshark log shows larger number of streams - 105 TCP streams for the same action that WinHTTP performs in 6 TCP streams. 

- I am looking for a solution to recude TCP Streams.

Curl stream sample:
![image](https://user-images.githubusercontent.com/2648522/112245075-3d746580-8c76-11eb-8dca-5c66323b775e.png)

WinHTTP stream sample:
![image](https://user-images.githubusercontent.com/2648522/112245144-58df7080-8c76-11eb-875e-df9ffe8f5bbf.png)


2) In Wireshark logs I am seeing more occurences of "TCP Retransmission", "TCP Dup Ack", "TCP previous segment was not captured", "TCP out-of-order".

- I am looking for a solution to avoid TCP related errors.

Sample TCP errors when using Curl:
Sample 1:
![image](https://user-images.githubusercontent.com/2648522/112245220-7a405c80-8c76-11eb-8285-51424b156b53.png)

Sample 2:
![image](https://user-images.githubusercontent.com/2648522/112245299-980dc180-8c76-11eb-847d-336a52c0bda8.png)

Sample 3:
![image](https://user-images.githubusercontent.com/2648522/112245349-b2e03600-8c76-11eb-9772-d77cf3a522c8.png)

Sample 4:
![image](https://user-images.githubusercontent.com/2648522/112245415-d86d3f80-8c76-11eb-8373-c6c0c053bdcb.png)


3) In Wireshark logs there is multiple occurence of "RST ACK" when connection fails.

- I don't know what is causing the reset acknowledge, but I want to prevent it.

Sample with RST ACK:
![image](https://user-images.githubusercontent.com/2648522/112245469-f33fb400-8c76-11eb-94a7-0e07dfed81ea.png)

**Reply by bagder** [reply]
You're not explaining what you do or how you do it, so we can't comment on if this is to be expected or not. The comparison with other tools is probably irrelevant since they're used completely different.

**Reply by jay** [reply]
I don't see anything unusual here. TCP is doing what it's designed to do. If you have a lot of transfers to the same host then you should reuse the easy handle, or if in a multi you can enable [CURLOPT_PIPEWAIT](https://curl.se/libcurl/c/CURLOPT_PIPEWAIT.html). Refer to [multi_vs_easy](https://gist.github.com/jay/97c8a7c20031997438dd7456e0a2b83a) test.

~~~
Iteration 5 of 5:

Testing 24 consecutive easy transfers to same host WITHOUT reusing handle.
It took 3.479 seconds to complete (easy) (reuse handle: NO)
(Multiple runs of this test averaged 3.710 seconds.)

Testing 24 consecutive easy transfers to same host WITH reusing handle.
It took 1.778 seconds to complete (easy) (reuse handle: YES)
(Multiple runs of this test averaged 2.156 seconds.)

Testing 24 concurrent easy transfers to same host from multi WITHOUT pipewait.
It took 0.561 seconds to complete (multi) (pipewait: NO)
(Multiple runs of this test averaged 0.542 seconds.)

Testing 24 concurrent easy transfers to same host from multi WITH pipewait.
It took 0.421 seconds to complete (multi) (pipewait: YES)
(Multiple runs of this test averaged 0.418 seconds.)
~~~

**Reply by ganeshkamath89** [reply]
Hi @jay,
     I ran your program. It gave me the following output:
libcurl/7.74.0-DEV OpenSSL/1.1.1i
Asynchronous DNS (threaded)



There will be a brief delay of 3000 ms after each test.



Iteration 1 of 5:

Testing 24 consecutive easy transfers to same host WITHOUT reusing handle.
It took 25.938 seconds to complete (easy) (reuse handle: NO)

Testing 24 consecutive easy transfers to same host WITH reusing handle.
It took 9.563 seconds to complete (easy) (reuse handle: YES)

Testing 24 concurrent easy transfers to same host from multi WITHOUT pipewait.
It took 1.375 seconds to complete (multi) (pipewait: NO)

Testing 24 concurrent easy transfers to same host from multi WITH pipewait.
It took 1.328 seconds to complete (multi) (pipewait: YES)



Iteration 2 of 5:

Testing 24 consecutive easy transfers to same host WITHOUT reusing handle.
It took 25.594 seconds to complete (easy) (reuse handle: NO)

Testing 24 consecutive easy transfers to same host WITH reusing handle.
It took 8.234 seconds to complete (easy) (reuse handle: YES)

Testing 24 concurrent easy transfers to same host from multi WITHOUT pipewait.
It took 1.328 seconds to complete (multi) (pipewait: NO)

Testing 24 concurrent easy transfers to same host from multi WITH pipewait.
It took 1.343 seconds to complete (multi) (pipewait: YES)



Iteration 3 of 5:

Testing 24 consecutive easy transfers to same host WITHOUT reusing handle.
It took 26.953 seconds to complete (easy) (reuse handle: NO)

Testing 24 consecutive easy transfers to same host WITH reusing handle.
It took 7.906 seconds to complete (easy) (reuse handle: YES)

Testing 24 concurrent easy transfers to same host from multi WITHOUT pipewait.
It took 1.328 seconds to complete (multi) (pipewait: NO)

Testing 24 concurrent easy transfers to same host from multi WITH pipewait.
It took 1.328 seconds to complete (multi) (pipewait: YES)



Iteration 4 of 5:

Testing 24 consecutive easy transfers to same host WITHOUT reusing handle.
It took 26.250 seconds to complete (easy) (reuse handle: NO)

Testing 24 consecutive easy transfers to same host WITH reusing handle.
It took 7.985 seconds to complete (easy) (reuse handle: YES)

Testing 24 concurrent easy transfers to same host from multi WITHOUT pipewait.
It took 1.328 seconds to complete (multi) (pipewait: NO)

Testing 24 concurrent easy transfers to same host from multi WITH pipewait.
It took 1.375 seconds to complete (multi) (pipewait: YES)



Iteration 5 of 5:

Testing 24 consecutive easy transfers to same host WITHOUT reusing handle.
It took 25.687 seconds to complete (easy) (reuse handle: NO)
(Multiple runs of this test averaged 26.084 seconds.)

Testing 24 consecutive easy transfers to same host WITH reusing handle.
It took 9.563 seconds to complete (easy) (reuse handle: YES)
(Multiple runs of this test averaged 8.650 seconds.)

Testing 24 concurrent easy transfers to same host from multi WITHOUT pipewait.
It took 1.359 seconds to complete (multi) (pipewait: NO)
(Multiple runs of this test averaged 1.343 seconds.)

Testing 24 concurrent easy transfers to same host from multi WITH pipewait.
It took 1.312 seconds to complete (multi) (pipewait: YES)
(Multiple runs of this test averaged 1.337 seconds.)

I am not so familiar with multi, need to study your code and check if multi works for us.
So you are using 
1) easy init with/without handel reuse
2) multi init with/without handle reuse - with/without pipewait

I think you can display more combinations.

**Reply by ganeshkamath89** [reply]
Hi @jay. How is handle reuse being done in your example, is it happening from this this line:

curl_easy_setopt(easy, CURLOPT_PRIVATE, user);

In general, if I use the curl opt header CURLOPT_PRIVATE and pass the URL in the third parameter, will it create a reusable handle, if the date is the same?

For example, if my URL is associated with this line:
`curl_easy_setopt(curl, CURLOPT_URL, cl->data.endpoint);`

if I add the header and associate it with the same value as URL like this:
```
curl_easy_setopt(curl, CURLOPT_URL, cl->data.endpoint);
curl_easy_setopt(curl, CURLOPT_PRIVATE, cl->data.endpoint);
```

 will it resuse the handle for future connections with the same URL?

Am I applying the code properly here?

**Reply by ganeshkamath89** [reply]
Hi @jay . Can you help me convert this file to easy with handle reuse?
File: https://github.com/Openwsman/openwsman/blob/master/src/lib/wsman-curl-client-transport.c

I don't know if we will be making multiple connections simultaneously to use multi with this file. I can try multi and check feasability with this file. But I don't know yet. Currently the handle reuse option seems promising.

I have added the PipeWait header "CURLOPT_PIPEWAIT" today, need to see how it affects performance. Thanks for this tip.

**Reply by jay** [reply]
> Am I applying the code properly here?

The code was written only for the purpose of comparison of the different methods. I have added that information to the comments and also I added some code to show errors.

> Can you help me convert this file to easy with handle reuse?

No. You would have to decide which method is appropriate for what you are doing and make those changes. The file you linked to is using curl_easy_perform. It is also possible to share connections in a single thread by using [CURL_LOCK_DATA_CONNECT](https://curl.se/libcurl/c/curl_share_setopt.html). For example if it's too much work to rewrite all that code to reuse the easy handle then you could reuse a shared object with connection sharing enabled instead. See example [shared-connection-cache.c](https://curl.se/libcurl/c/shared-connection-cache.html).

**Reply by ganeshkamath89** [reply]
Hi @jay.
     We have observed that over VPN curl connection breaks. The pattern shared in the beginning of the discussion is seen more frequently on VPN based connections. Do you have any recommendations what I can do to establish the connection, longer or enable it to be more persistant? Do you have any recommendations for any particular Curl_Opt headers or other settings?

**Reply by jay** [reply]
Sorry no, it sounds like network issues and not a curl problem.
